kpatch-build/create-diff-object.c

```diff
@@ -771,6 +771,88 @@ static bool kpatch_line_macro_change_only(struct section *sec)
 
 	return true;
 }
+
+#elif __s390x__
+
+static int _insn_get_length(unsigned char *inst)
+{
+	int length;
+
+	switch (inst[0] >> 6) {
+		case 0b00:
+			length = 2;
+			break;
+		case 0b01:
+		case 0b10:
+			length = 4;
+			break;
+		case 0b11:
+			length = 6;
+			break;
+	}
+
+	return length;
+}
+
+static bool kpatch_line_macro_change_only(struct section *sec)
+{
+	unsigned char *start1, *start2;
+	unsigned long size, offset, length;
+	struct rela *rela;
+	int lineonly = 0, found;
+
+	if (sec->status != CHANGED ||
+	    is_rela_section(sec) ||
+	    !is_text_section(sec) ||
+	    sec->sh.sh_size != sec->twin->sh.sh_size ||
+	    !sec->rela ||
+	    sec->rela->status != SAME)
+		return false;
+
+	start1 = sec->twin->data->d_buf;
+	start2 = sec->data->d_buf;
+	size = sec->sh.sh_size;
+	for (offset = 0; offset < size; offset += length) {
+		length = _insn_get_length(start1 + offset);
+
+		if (_insn_get_length(start1 + offset) != _insn_get_length(start2 + offset))
+			return false;
+
+		if (!memcmp(start1 + offset, start2 + offset, length))
+			continue;
+
+		if ((start1[offset] != 0xa7) || (start1[offset+1] != 0x49) ||
+		    (start2[offset] != 0xa7) || (start2[offset+1] != 0x49))
+			return false;
+
+		/*
+		 * Verify zero or more string relas followed by a
+		 * warn_slowpath_* or another similar rela.
+		 */
+		found = 0;
+		list_for_each_entry(rela, &sec->rela->relas, list) {
```

Comment:
For the other arches, we look for a relocation to static local variable `__warned` (and soon `__already_done` per #1241).  In this commit msg's disassembly listing, the compiled order throughout the object file was:

1. `R_390_PC32DBL -> .data.once+0x2` is static local  `__warned.xxxx` or `__already_done.xxxxx`
2. `lghi %r4, <line number>` 
3. `R_390_PC32DBL -> .LC1+0x2` is a string rela
4. `R_390_PC32DBL -> .LC2+0x2` is a string rela
5. `R_390_PLT32DBL-> __warn_printk+0x2` is function rela

When testing this change, @C-Erastus and I only focused on pattern parts 2-5 above.  Can we make any assumptions about the relocation ordering, or should we add back checks for the static local variables just in case they land after the `lghi` instruction?

Reply:
I have no knowledge of whether this ordering can be guaranteed but it seems to me that having the check 'just in case' has more upsides than downsides.